libkatana_python_native/include/katana/python/NumbaSupport.h

```diff
@@ -14,138 +14,207 @@ namespace detail {
+
+/// A utility class encapsulating the type manipulation required for binding
+/// non-const member functions.
+template <typename _Cls, typename _Return, typename... _Args>
+struct MemberFunctionBinderImpl {
+  using Cls = _Cls;
+  using Return = _Return;
+  template <typename Impl>
+  using BaseMemberFunctionPointer = Return (Impl::*)(_Args...);
+  template <typename Impl>
+  using BaseStaticFunctionPointer = Return (*)(Impl*, remove_cvref_t<_Args>...);
+  using MemberFunctionPointer = BaseMemberFunctionPointer<Cls>;
+  using StaticFunctionPointer = BaseStaticFunctionPointer<Cls>;
+
+  /// A static wrapper function which forwards calls to the statically provided
+  /// member function pointer.
+  template <MemberFunctionPointer member_func>
+  static remove_cvref_t<Return> call(Cls* self, remove_cvref_t<_Args>... args) {
+    return (self->*(member_func))(args...);
   }
-};
 
-template <typename... Args>
-class DefWithNumbaImpl {
-  /// Define an actual python method and numba wrapper using the provided
-  /// function pointer and data pointer (@p caller).
+  /// Define an actual python method and numba wrapper using @p member_func.
   template <
-      typename Cls, typename Return, typename Func, typename Caller,
-      typename... Extra>
-  void def_class_method(
-      pybind11::class_<Cls>& cls, const char* name, Func f, Caller* caller,
-      const Extra&... extra) const {
-    cls.def(name, f, extra...);
+      MemberFunctionPointer member_func, typename SelfCls, typename... Extra>
+  static void def_class_method(
+      pybind11::class_<SelfCls>& cls, const char* name, const Extra&... extra) {
+    cls.def(name, member_func, extra...);
     auto numba_support =
         pybind11::module_::import("katana.native_interfacing.numba_support");
     numba_support.attr("register_method")(
-        cls, cls.attr(name), (uintptr_t)&Caller::call, (uintptr_t)caller,
+        cls, cls.attr(name), (uintptr_t)&call<member_func>,
         PythonTypeTraits<remove_cvref_t<Return>>::ctypes_type(),
-        PythonTypeTraits<remove_cvref_t<Args>>::ctypes_type()...);
+        PythonTypeTraits<remove_cvref_t<_Args>>::ctypes_type()...);
   }
 
-  /// Define an actual python funct8ion and numba wrapper using the provided
-  /// function pointer. Unlike methods, there is no need for a data pointer.
-  template <typename Return, typename... Extra>
-  void def_func(
-      pybind11::module_& m, const char* name, Return (*f)(Args...),
-      const Extra&... extra) const {
-    m.def(name, f, extra...);
-    auto func = m.attr(name);
+  constexpr static bool is_const = false;
+};
+
+/// A utility class encapsulating the type manipulation required for binding
+/// const member functions.
+template <typename _Cls, typename _Return, typename... _Args>
+struct ConstMemberFunctionBinderImpl {
+  using Cls = _Cls;
+  using Return = _Return;
+  template <typename Impl>
+  using BaseMemberFunctionPointer =
+      Return (Impl::*)(remove_cvref_t<_Args>...) const;
```

Comment:
What's going on here?

Reply:
I'm defining a type member which is a member function pointer with a configurable class and has the same return type, but has arguments which are modified by removing `const` `volatile` and `&` (reference). The configurable class is needed to handle methods that are defined on a base class, but are being used on subclass instances.